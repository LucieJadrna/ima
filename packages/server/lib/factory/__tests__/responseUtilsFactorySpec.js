'use strict';

const responseUtilsFactory = require('../responseUtilsFactory.js');
const manifestMock = require('../__mocks__/manifest.json');

jest.mock('fs', () => {
  const { toMockedInstance } = jest.requireActual('to-mock');
  const originalModule = jest.requireActual('fs');

  return {
    ...toMockedInstance(originalModule, {
      existsSync() {
        return true;
      },
      readFileSync(path) {
        if (path.endsWith('manifest.json')) {
          return JSON.stringify(manifestMock);
        }

        return 'runner#{source}';
      },
    }),
  };
});

describe('responseUtilsFactory', () => {
  const {
    processContent,
    createContentVariables,
    _renderStyles,
    _prepareCookieOptionsForExpress,
    _prepareSource,
  } = responseUtilsFactory();

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('_renderStyles', () => {
    it('should return empty string for invalid input', () => {
      expect(_renderStyles()).toBe('');
      expect(_renderStyles([])).toBe('');
      expect(_renderStyles({})).toBe('');
      expect(_renderStyles(null)).toBe('');
      expect(_renderStyles('/static/app.css')).toBe('');
    });

    it('should return link stylesheet tags for string items', () => {
      expect(_renderStyles(['/static/app.css'])).toBe(
        '<link rel="stylesheet" href="/static/app.css" />'
      );

      expect(
        _renderStyles([
          '/static/app1.css',
          '/static/app2.css',
          '/static/app3.css',
        ])
      ).toBe(
        '<link rel="stylesheet" href="/static/app1.css" />' +
          '<link rel="stylesheet" href="/static/app2.css" />' +
          '<link rel="stylesheet" href="/static/app3.css" />'
      );
    });

    it('should return link tag with custom attributes', () => {
      expect(
        _renderStyles([
          [
            '/static/app.css',
            { type: 'text/css', rel: 'preload', as: 'style' },
          ],
        ])
      ).toBe(
        '<link href="/static/app.css" rel="preload" type="text/css" as="style" />'
      );
    });

    it('should insert custom mini script for fallback sources', () => {
      expect(
        _renderStyles([
          [
            '/static/app.css',
            { rel: 'stylesheet', fallback: '/static/fallback.css' },
          ],
        ])
      ).toBe(
        `<link href="/static/app.css" rel="stylesheet" onerror="this.onerror=null;this.href='/static/fallback.css';" />`
      );
    });
  });

  describe('_prepareSource', () => {
    it('should prepare default sources structure from provided manifest file', () => {
      expect(_prepareSource(manifestMock, 'en')).toMatchSnapshot();
    });

    it('should add fallbacks when CDN_STATIC_ROOT_URL is defined', () => {
      process.env.CDN_STATIC_ROOT_URL = 'cdn://';

      expect(_prepareSource(manifestMock, 'en')).toMatchSnapshot();
      process.env.CDN_STATIC_ROOT_URL = '';
    });

    it('should skip compilations without assets', () => {
      const sources = _prepareSource(
        {
          ...manifestMock,
          assetsByCompiler: {
            ...manifestMock.assetsByCompiler,
            client: {},
          },
        },
        'en'
      );

      expect(sources).toMatchSnapshot();
    });
  });

  describe('_prepareCookieOptionsForExpress', () => {
    it('should convert cookie maxAge to ms for Express', () => {
      let options = { maxAge: 1 };
      let expressOptions = _prepareCookieOptionsForExpress(options);
      expect(options.maxAge).toBe(1);
      expect(expressOptions.maxAge).toBe(1000);
    });

    it('should remove cookie maxAge: null for Express', () => {
      // Because Express converts null to 0, which is not intended.
      let options = { maxAge: null };
      let expressOptions = _prepareCookieOptionsForExpress(options);
      expect(options.maxAge).toBeNull();
      expect(expressOptions.maxAge).toBeUndefined();
    });
  });

  describe('createContentVariables', () => {
    it('should return empty object if there is no valid bootConfig', () => {
      expect(createContentVariables({})).toStrictEqual({});
      expect(createContentVariables({ bootConfig: null })).toStrictEqual({});
      expect(createContentVariables({ bootConfig: {} })).toStrictEqual({});
      expect(
        createContentVariables({ bootConfig: { settings: null } })
      ).toStrictEqual({});
    });

    it('should generate base set of content variables', () => {
      const response = {
        content: '<html>#{styles}#{revivalSettings}#{runner}</html>',
      };
      const bootConfig = {
        settings: {
          $Language: 'en',
          $Debug: true,
        },
      };

      expect(
        createContentVariables({ bootConfig, response })
      ).toMatchSnapshot();
    });
  });

  describe('processContent', () => {
    it('should return original content without any boot config', () => {
      expect(processContent({ response: { content: 'content' } })).toBe(
        'content'
      );
    });

    it('should interpolate revival scripts into page content', () => {
      const bootConfig = {
        settings: {
          $Language: 'en',
          $Debug: true,
        },
      };
      const response = {
        content: '<html>#{styles}#{revivalSettings}#{runner}</html>',
        contentVariables: createContentVariables({ bootConfig, response: {} }),
      };
      const contextMock = { response, bootConfig };

      const content = processContent(contextMock);
      expect(content).toMatchSnapshot();
    });

    it('should allow overrides through custom $Source definition', () => {
      const bootConfig = {
        settings: {
          $Language: 'en',
          $Debug: true,
          $Source: (context, manifest, sources) => {
            return {
              styles: [],
              esScripts: [...sources.scripts, 'custom-script-src'],
            };
          },
        },
      };
      const response = {
        content: '<html>#{styles}#{revivalSettings}#{runner}</html>',
        contentVariables: createContentVariables({ bootConfig, response: {} }),
      };
      const contextMock = { response, bootConfig };
      const content = processContent(contextMock);

      expect(content).toMatchSnapshot();
    });

    it('should interpolate deep embedded variables', () => {
      const bootConfig = {
        settings: {
          $Language: 'en',
          $Debug: true,
        },
      };
      const response = {
        content: '<html>#{v1}</html>',
        contentVariables: {
          v1: '#{v2}',
          v2: '#{v3}',
          v3: '#{v4}',
          v4: '#{v5}',
          v5: '#{v6}',
          v6: '#{v7}',
          v7: 'final v7 content',
        },
      };
      const contextMock = { response, bootConfig };
      const content = processContent(contextMock);

      expect(content).toBe('<html>final v7 content</html>');
    });

    it('should allow interpolation of settings variables from bootConfig', () => {
      const bootConfig = {
        settings: {
          $Language: 'en',
          $Version: '1.0.0',
          $Debug: true,
        },
      };
      const response = {
        content: '<html>#{c1}: #{$Version}</html>',
        contentVariables: {
          c1: 'content-variable',
        },
      };
      const contextMock = { response, bootConfig };
      const content = processContent(contextMock);

      expect(content).toBe('<html>content-variable: 1.0.0</html>');
    });
  });
});
