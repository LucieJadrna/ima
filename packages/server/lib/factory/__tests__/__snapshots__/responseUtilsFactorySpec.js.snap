// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`responseUtilsFactory _prepareSource should add fallbacks when CDN_STATIC_ROOT_URL is defined 1`] = `
{
  "esScripts": [
    [
      "cdn://static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js.es/app.client.2106a34c6b8bbad8.js",
      },
    ],
    [
      "cdn://static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js.es/locale/en.371127fdbfbe93d2.js",
      },
    ],
    [
      "cdn://static/js.es/vendors.a873907f25297544.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js.es/vendors.a873907f25297544.js",
      },
    ],
  ],
  "scripts": [
    [
      "cdn://static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js/app.client.8dc14fd2c9e52eef.js",
      },
    ],
    [
      "cdn://static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      },
    ],
    [
      "cdn://static/js/vendors.0e456297851f0a3a.js",
      {
        "async": true,
        "crossorigin": "anonymous",
        "fallback": "/static/js/vendors.0e456297851f0a3a.js",
      },
    ],
  ],
  "styles": [
    [
      "cdn://static/css/app.d0ad44d05f82db5f.css",
      {
        "fallback": "/static/css/app.d0ad44d05f82db5f.css",
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
}
`;

exports[`responseUtilsFactory _prepareSource should prepare default sources structure from provided manifest file 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
  ],
  "scripts": [
    [
      "/static/js/app.client.8dc14fd2c9e52eef.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/locale/en.3c6a5e7a55bb2ab4.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js/vendors.0e456297851f0a3a.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
  ],
  "styles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
}
`;

exports[`responseUtilsFactory _prepareSource should skip compilations without assets 1`] = `
{
  "esScripts": [
    [
      "/static/js.es/app.client.2106a34c6b8bbad8.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/locale/en.371127fdbfbe93d2.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
    [
      "/static/js.es/vendors.a873907f25297544.js",
      {
        "async": true,
        "crossorigin": "anonymous",
      },
    ],
  ],
  "scripts": [],
  "styles": [
    [
      "/static/css/app.d0ad44d05f82db5f.css",
      {
        "preload": true,
        "rel": "stylesheet",
      },
    ],
  ],
}
`;

exports[`responseUtilsFactory processContent should allow overrides through custom $Source definition 1`] = `
"<html><script id="ima-revival-settings">(function (root) {
      root.$Debug = true;
      root.$IMA = root.$IMA || {};
      $IMA.SPA = false;
      $IMA.$Language = "en";
      $IMA.$Env = "undefined";
      $IMA.$Debug = true;
      $IMA.$Version = "undefined";
      $IMA.$App = undefined;
      $IMA.$Protocol = "undefined";
      $IMA.$Host = "undefined";
      $IMA.$Path = "undefined";
      $IMA.$Root = "undefined";
      $IMA.$LanguagePartPath = "undefined";
    })(typeof window !== 'undefined' && window !== null ? window : global);
    </script><script id="ima-runner">runner{\\"esScripts\\":[[\\"/static/js/app.client.8dc14fd2c9e52eef.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js/locale/en.3c6a5e7a55bb2ab4.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js/vendors.0e456297851f0a3a.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],\\"custom-script-src\\"]}</script><script id="ima-revival-cache">(function (root) {
      root.$IMA = root.$IMA || {};
      $IMA.Cache = {};
    })(typeof window !== 'undefined' && window !== null ? window : global);
    </script></html>"
`;

exports[`responseUtilsFactory processContent should interpolate revival scripts into page content 1`] = `
"<html><link href="/static/css/app.d0ad44d05f82db5f.css" rel="stylesheet" /><script id="ima-revival-settings">(function (root) {
      root.$Debug = true;
      root.$IMA = root.$IMA || {};
      $IMA.SPA = false;
      $IMA.$Language = "en";
      $IMA.$Env = "undefined";
      $IMA.$Debug = true;
      $IMA.$Version = "undefined";
      $IMA.$App = undefined;
      $IMA.$Protocol = "undefined";
      $IMA.$Host = "undefined";
      $IMA.$Path = "undefined";
      $IMA.$Root = "undefined";
      $IMA.$LanguagePartPath = "undefined";
    })(typeof window !== 'undefined' && window !== null ? window : global);
    </script><script id="ima-runner">runner{\\"scripts\\":[[\\"/static/js/app.client.8dc14fd2c9e52eef.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js/locale/en.3c6a5e7a55bb2ab4.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js/vendors.0e456297851f0a3a.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}]],\\"esScripts\\":[[\\"/static/js.es/app.client.2106a34c6b8bbad8.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js.es/locale/en.371127fdbfbe93d2.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}],[\\"/static/js.es/vendors.a873907f25297544.js\\",{\\"async\\":true,\\"crossorigin\\":\\"anonymous\\"}]]}</script></html>"
`;
